op-chain-ops/cmd/op-run-block/main.go

```diff
@@ -0,0 +1,363 @@
+	}
+
+	block, err := loadBlock(blockPath)
+	if err != nil {
+		return fmt.Errorf("failed to load block: %w", err)
+	}
+	if err := block.Verify(); err != nil {
+		return fmt.Errorf("block content is invalid: %w", err)
+	}
+	logger.Info("Loaded block",
+		"hash", block.Hash, "number", uint64(block.Number), "txs", len(block.Transactions))
+
+	parentBlock, err := ethCl.HeaderByHash(ctx, block.ParentHash)
+	if err != nil {
+		return fmt.Errorf("failed to fetch parent block: %w", err)
+	}
+
+	stateDB := gstate.NewDatabase(triedb.NewDatabase(db, &triedb.Config{
+		Preimages: true,
+	}), nil)
+	state, err := gstate.New(parentBlock.Root, stateDB)
+	if err != nil {
+		return fmt.Errorf("failed to create in-memory state: %w", err)
+	}
+
+	header := block.RPCHeader.CreateGethHeader()
+
+	vmCfg := vm.Config{Tracer: nil}
+	consensusEng := beacon.New(&beacon.OpLegacy{})
+	chCtx := &remoteChainCtx{
+		consensusEng: consensusEng,
+		hdr:          header,
+		cfg:          config,
+		cl:           ethCl,
+		logger:       logger,
+	}
+
+	outW, err := os.OpenFile("tx_17_dump.json", os.O_CREATE|os.O_WRONLY, 0755)
+	if err != nil {
+		return fmt.Errorf("failed to create/open dump file: %w", err)
+	}
+	defer outW.Close()
+
+	vmCfg.Tracer = logger2.NewJSONLogger(&logger2.Config{
+		EnableMemory:     false,
+		DisableStack:     false,
+		DisableStorage:   false,
+		EnableReturnData: false,
+		Debug:            false,
+		Limit:            0,
+		Overrides:        nil,
+	}, outW)
+
+	witness, err := stateless.NewWitness(header, chCtx)
+	if err != nil {
+		return fmt.Errorf("failed to prepare witness data collector: %w", err)
+	}
+	state.StartPrefetcher("debug", witness)
+	defer func() {
+		witnessDump := witness.ToExecutionWitness()
+		out, err := json.MarshalIndent(witnessDump, "", "  ")
+		if err != nil {
+			logger.Error("failed to encode witness", "err", err)
+			return
+		}
+		if err := os.WriteFile("debug_witness.json", out, 0755); err != nil {
+			logger.Error("Failed to write witness", "err", err)
+		}
+	}()
+	logger.Info("Starting block processing")
+	result, err := Process(logger, config, block, state, vmCfg, chCtx, outW)
+	if err != nil {
+		return fmt.Errorf("failed to process: %w", err)
+	}
+	logger.Info("Done", "gas_used", result.GasUsed)
+	return nil
+}
+
+type WrappedDump struct {
+	Block *sources.RPCBlock `json:"block"`
+}
+
+func loadBlock(blockPath string) (*sources.RPCBlock, error) {
+	data, err := os.ReadFile(blockPath)
+	if err != nil {
+		return nil, err
+	}
+	var out []WrappedDump
+	if err := json.Unmarshal(data, &out); err != nil {
+		return nil, err
+	}
+	if len(out) != 1 {
+		return nil, fmt.Errorf("expected single block entry, got %d", len(out))
+	}
+	blockData := out[0].Block
+	return blockData, nil
+}
+
+type remoteChainCtx struct {
+	consensusEng consensus.Engine
+	hdr          *types.Header
+	cfg          *params.ChainConfig
+	cl           *ethclient.Client
+	logger       log.Logger
+}
```

Comment:
Would this structure be a good candidate for promotion to a shared package space like `op-service`?

Reply:
Maybe, I think it's not flexible enough, and a bit inefficient, to really become a shared op-service util.